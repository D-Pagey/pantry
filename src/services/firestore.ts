import { BatchType, DatabaseFoodType } from '../types';
import { db, firebase } from '.';

type UpdateBatchProps = {
    name: string;
    batch: BatchType;
    userHousehold: string;
};

/**
 * This function updates a specific batch within an existing item
 */
export const updateBatch = ({ name, batch, userHousehold }: UpdateBatchProps): Promise<void> => {
    return db
        .collection('households')
        .doc(userHousehold)
        .update({
            [`fridge.${name}.batches.${batch.id}`]: batch
        });
};

/**
 * This function deletes all batches for an item
 */
export const deleteItemBatches = (name: string, household: string): Promise<void> => {
    return db
        .collection('households')
        .doc(household)
        .update({
            [`fridge.${name}.batches`]: {}
        });
};

/**
 * This function adds a brand new item or overwrites an existing one
 */
export const addItem = (newItem: DatabaseFoodType, household: string): Promise<void> => {
    if (newItem.name.split('').includes('.')) {
        throw new Error('Item name should not contain a period');
    }

    return db
        .collection('households')
        .doc(household)
        .update({
            [`fridge.${newItem.name}`]: newItem
        });
};

/**
 * This function adds a brand new item or overwrites an existing one
 */
export const addNewUnit = (units: string[], household: string): Promise<void> => {
    return db
        .collection('households')
        .doc(household)
        .update({
            [`meta.units`]: units
        });
};

/**
 * This function adds a brand new item then deletes the old item out of firestore
 */
export const addItemDeleteItem = async (
    newItem: DatabaseFoodType,
    nameToBeDeleted: string,
    household: string
): Promise<void> => {
    // add new item
    await addItem(newItem, household);

    // delete old item
    return deleteItemBatches(nameToBeDeleted, household);
};

type DeleteBatchProps = {
    name: string;
    batchId: string;
    userHousehold: string;
};

/**
 * This function deletes a specific batchId from firestore
 */
export const deleteBatch = ({ name, batchId, userHousehold }: DeleteBatchProps): Promise<void> => {
    return db
        .collection('households')
        .doc(userHousehold)
        .update({
            [`fridge.${name}.batches.${batchId}`]: firebase.firestore.FieldValue.delete()
        });
};
