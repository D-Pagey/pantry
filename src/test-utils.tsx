import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import React, { FC, ReactNode } from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { FirebaseContext } from './components/ProviderFirebase';

type AllTheProvidersProps = {
    children: ReactNode;
    customContext?: any;
};

const AllTheProviders: FC<AllTheProvidersProps> = ({ children, customContext }) => (
    <FirebaseContext.Provider
        value={{
            categories: [],
            deleteFoodItem: () => () => null,
            expiringCount: 0,
            isAuthed: false,
            isCheckingAuth: false,
            fridge: [],
            signOut: () => null,
            updateFridge: () => null,
            user: {
                email: null,
                name: null
            },
            ...customContext
        }}
    >
        <MemoryRouter>{children}</MemoryRouter>
    </FirebaseContext.Provider>
);

const customRender = (ui: any, customContext?: any, options?: any) =>
    render(ui, { wrapper: () => <AllTheProviders customContext={customContext}>{ui}</AllTheProviders>, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
