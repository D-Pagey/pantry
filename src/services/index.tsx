import fb from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';
import 'firebase/messaging';

const firebaseConfig = {
    apiKey: process.env.REACT_APP_API_KEY,
    authDomain: process.env.REACT_APP_AUTH_DOMAIN,
    databaseURL: process.env.REACT_APP_DATABASE_URL,
    projectId: process.env.REACT_APP_PROJECT_ID,
    storageBucket: process.env.REACT_APP_STORAGE_BUCKET,
    messagingSenderId: process.env.REACT_APP_MESSAGING_SENDER_ID,
    appId: process.env.REACT_APP_FIREBASE_APP_ID
};

// Initialize Cloud Firestore through Firebase
fb.initializeApp(firebaseConfig);

export const firebase = fb;

const messaging = fb.messaging();

// Add the public key generated from the console here.
messaging.usePublicVapidKey(
    'BACbATPdC2mrqm7mSvlv9xln-fXSs9-qZEoNgS7svsXlRoPYt7vq7T7asD0C2YwXKZ0dLukM5d2KQolWWZtAJNE'
);

// Get Instance ID token. Initially this makes a network call, once retrieved
// subsequent calls to getToken will return from cache.
messaging
    .getToken()
    .then((currentToken) => {
        if (currentToken) {
            console.log({ currentToken });
        } else {
            // Show permission request.
            console.log('No Instance ID token available. Request permission to generate one.');
            // Show permission UI.
        }
    })
    .catch((err) => {
        console.log('An error occurred while retrieving token. ', err);
    });

messaging.onMessage((payload) => console.log({ config: payload }));

// Callback fired if Instance ID token is updated.
messaging.onTokenRefresh(() => {
    messaging
        .getToken()
        .then((refreshedToken) => {
            console.log('Token refreshed.', refreshedToken);
            // Indicate that the new Instance ID token has not yet been sent to the
            // app server.

            // Send Instance ID token to app server.

            // ...
        })
        .catch((err) => {
            console.log('Unable to retrieve refreshed token ', err);
        });
});
