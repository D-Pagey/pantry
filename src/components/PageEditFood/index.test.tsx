import React from 'react';
import selectEvent from 'react-select-event';
import { render, screen } from '../../test-utils';
import { Fridge, TenantHeidi, UserDan } from '../../fixtures';
import { addItemDeleteItem, updateItemField } from '../../services/firestore';
import { PageEditFood } from '.';
import userEvent from '@testing-library/user-event';

jest.mock('../../services/firestore');
jest.mock('react-router-dom', () => ({
    // @ts-ignore
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({
        name: 'broccoli'
    })
}));

const props = {
    fridge: Fridge,
    tenants: [TenantHeidi]
};

const context = {
    user: UserDan
};

const batchesObject = {
    '1111111': {
        expires: expect.any(Date),
        id: '1111111',
        ownerId: 'fghij',
        quantity: 1
    },
    '22222222': {
        expires: expect.any(Date),
        id: '22222222',
        ownerId: 'abcde',
        quantity: 2
    },
    '3333333': {
        expires: expect.any(Date),
        id: '3333333',
        ownerId: 'zxwy',
        quantity: 4
    }
};

describe('PageEditFood component', () => {
    it('should render', () => {
        const { container } = render(<PageEditFood {...props} />, context);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should call addItemDeleteItem when name changes to a new name', async () => {
        const newName = 'editing-item-name';

        render(<PageEditFood {...props} />, context);

        await selectEvent.create(screen.getByLabelText('Change item name:'), newName);

        userEvent.click(screen.getByText('Save Changes'));

        expect(addItemDeleteItem).toHaveBeenCalledWith(
            {
                batches: batchesObject,
                category: 'vegetables',
                name: newName,
                unit: 'servings'
            },
            'broccoli',
            UserDan.household
        );
    });

    it('should call addItemDeleteItem when name changes to an existing name', async () => {
        render(<PageEditFood {...props} />, context);

        await selectEvent.select(screen.getByLabelText('Change item name:'), 'Steak (7 servings)');

        userEvent.click(screen.getByText('Save Changes'));

        expect(addItemDeleteItem).toHaveBeenCalledWith(
            {
                batches: batchesObject,
                category: 'meat',
                name: 'steak',
                unit: 'servings'
            },
            'broccoli',
            UserDan.household
        );
    });

    it('should call updateItemField if just the category changes', () => {
        const category = 'Dairy';

        render(<PageEditFood {...props} />, context);

        userEvent.click(screen.getByText(category));

        userEvent.click(screen.getByText('Save Changes'));

        expect(updateItemField).toHaveBeenCalledWith('broccoli', 'category', category.toLowerCase(), UserDan.household);
    });

    it('should call addItemDeleteItem when name and category changes to a new item name', async () => {
        const newName = 'new-name-and-category';
        const category = 'Dairy';

        render(<PageEditFood {...props} />, context);

        await selectEvent.create(screen.getByLabelText('Change item name:'), newName);
        userEvent.click(screen.getByText(category));

        userEvent.click(screen.getByText('Save Changes'));

        expect(addItemDeleteItem).toHaveBeenCalledWith(
            {
                batches: batchesObject,
                category: category.toLowerCase(),
                name: newName,
                unit: 'servings'
            },
            'broccoli',
            UserDan.household
        );
    });

    it('should call addItemDeleteItem when name and category changes to an existing name', async () => {
        const category = 'Dairy';

        render(<PageEditFood {...props} />, context);

        await selectEvent.select(screen.getByLabelText('Change item name:'), 'Steak (7 servings)');
        userEvent.click(screen.getByText(category));

        userEvent.click(screen.getByText('Save Changes'));

        expect(addItemDeleteItem).toHaveBeenCalledWith(
            {
                batches: batchesObject,
                category: category.toLowerCase(),
                name: 'steak',
                unit: 'servings'
            },
            'broccoli',
            UserDan.household
        );
    });
});
