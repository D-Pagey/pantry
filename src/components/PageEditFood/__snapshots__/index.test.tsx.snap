// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageEditFood component should render 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: white;
  box-shadow: 0 4px 8px 0 rgba(0,0,0,0.06);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 0 24px;
  left: 0;
  position: fixed;
  right: 0;
  top: 0;
  z-index: 2;
}

.c3 {
  font-size: 25px;
  font-weight: 500;
  margin: 0;
  padding: 0 0 0 14px;
}

.c2 {
  cursor: pointer;
  height: 16px;
  width: 16px;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="arrow"
      class="c2"
      data-testid="headerBackArrow"
      src="arrow.svg"
    />
    <h4
      class="c3"
    >
      Edit 
    </h4>
  </div>
</div>
`;
